frame/election-provider-multi-phase/src/lib.rs

```diff
@@ -1287,12 +1287,29 @@ impl<T: Config> Pallet<T> {
 		}
 
 		// Only write snapshot if all existed.
-		<SnapshotMetadata<T>>::put(SolutionOrSnapshotSize {
+		let metadata = SolutionOrSnapshotSize {
 			voters: voters.len() as u32,
 			targets: targets.len() as u32,
-		});
+		};
+		log!(debug, "creating a snapshot with metadata {:?}", metadata);
+
+		<SnapshotMetadata<T>>::put(metadata);
 		<DesiredTargets<T>>::put(desired_targets);
-		<Snapshot<T>>::put(RoundSnapshot { voters, targets });
+
+		// instead of using storage APIs, we do a manual encoding into a fixed-size buffer.
+		// encode it without storing it anywhere, this should not cause any allocation.
+		let snapshot = RoundSnapshot { voters, targets };
+		let size = snapshot.encoded_size();
+		log!(info, "snapshot pre-calculated size {:?}", size);
+		let mut buffer = Vec::with_capacity(size);
+		snapshot.encode_to(&mut buffer);
+
+		// do some checks.
+		debug_assert_eq!(buffer, snapshot.encode());
+		// buffer should have not re-allocated since.
+		debug_assert_eq!(buffer.len(), size);
```

Comment:
Shouldn't this be `buffer.capacity()`, if the goal is to ensure no further reallocation?

Reply:
it'll be more correct I presume, although my intention is to ensure `.len == .capacity == size`. 

I think this would be similar to implementing `codec::Output` for `BoundedVec`

Reply:
But you don't ensure that the capacity == len